synapse/api/auth.py

```diff
@@ -944,3 +961,175 @@ async def check_user_in_room_or_world_readable(
+        ), "Invalid client_auth_method"
+
+        self._http_client = hs.get_proxied_http_client()
+        self._hostname = hs.hostname
+
+        self._issuer_metadata = RetryOnExceptionCachedCall(self._load_metadata)
+        secret = self._config.oauth_delegation_client_secret
+        self._client_auth = ClientAuth(
+            self._config.oauth_delegation_client_id,
+            secret,
+            OAuthDelegatedAuth.AUTH_METHODS[
+                self._config.oauth_delegation_client_auth_method
+            ],
+        )
+
+    async def _load_metadata(self) -> OpenIDProviderMetadata:
+        if self._config.oauth_delegation_issuer_metadata is not None:
+            return OpenIDProviderMetadata(
+                **self._config.oauth_delegation_issuer_metadata
+            )
+        url = get_well_known_url(self._config.oauth_delegation_issuer, external=True)
+        response = await self._http_client.get_json(url)
+        metadata = OpenIDProviderMetadata(**response)
+        # metadata.validate_introspection_endpoint()
+        return metadata
+
+    async def _introspect_token(self, token: str) -> IntrospectionToken:
+        metadata = await self._issuer_metadata.get()
+        introspection_endpoint = metadata.get("introspection_endpoint")
+        raw_headers: Dict[str, str] = {
+            "Content-Type": "application/x-www-form-urlencoded",
+            "User-Agent": str(self._http_client.user_agent, "utf-8"),
+            "Accept": "application/json",
+        }
+
+        args = {"token": token, "token_type_hint": "access_token"}
+        body = urlencode(args, True)
+
+        # Fill the body/headers with credentials
+        uri, raw_headers, body = self._client_auth.prepare(
+            method="POST", uri=introspection_endpoint, headers=raw_headers, body=body
+        )
+        headers = Headers({k: [v] for (k, v) in raw_headers.items()})
+
+        # Do the actual request
+        # We're not using the SimpleHttpClient util methods as we don't want to
+        # check the HTTP status code and we do the body encoding ourself.
+        response = await self._http_client.request(
+            method="POST",
+            uri=uri,
+            data=body.encode("utf-8"),
+            headers=headers,
+        )
+
+        resp_body = await make_deferred_yieldable(readBody(response))
+        # TODO: Let's not worry about 5xx errors & co. for now and just try
+        # decoding that as JSON. We should also do some validation of the
+        # response
+        resp = json_decoder.decode(resp_body.decode("utf-8"))
+        return IntrospectionToken(**resp)
+
+    async def get_user_by_req(
+        self,
+        request: SynapseRequest,
+        allow_guest: bool = False,
+        allow_expired: bool = False,
+    ) -> Requester:
+        access_token = self.get_access_token_from_request(request)
+        return await self.get_user_by_access_token(access_token, allow_expired)
```

Comment:
I did this because in the `LegacyAuth` one it does a lot more things, like tracking user IPs, managing whether a guest is allowed to do the request, and includes the logic to handle ASes requests. We probably want to refactor this down the line, but in this PR I tried to avoid touching the existing auth logic

Reply:
OK! Sounds like we should add a TODO comment at least?